src/hotspot/share/classfile/loaderConstraints.cpp

```diff
@@ -185,62 +191,54 @@ class PurgeUnloadedConstraints : public StackObj {
-        probe->set_klass(NULL);
+      LoaderConstraint* constraint = set.constraint_at(i);
+      InstanceKlass* klass = constraint->klass();
+
+      // Remove klass and constraint that is no longer alive
+      if (klass != NULL && !klass->is_loader_alive()) {
         if (lt.is_enabled()) {
           ResourceMark rm;
           lt.print("purging class object from constraint for name %s,"
                      " loader list:",
                      name->as_C_string());
-          for (int i = 0; i < probe->num_loaders(); i++) {
+          for (int i = 0; i < constraint->num_loaders(); i++) {
             lt.print("    [%d]: %s", i,
-                          probe->loader_data(i)->loader_name_and_id());
+                          constraint->loader_data(i)->loader_name_and_id());
           }
         }
-      }
-
-      // Remove entries no longer alive from loader array
-      for (int n = probe->num_loaders() - 1; n >= 0; n--) {
-        if (probe->loader_data(n)->is_unloading()) {
-          if (lt.is_enabled()) {
-            ResourceMark rm;
-            lt.print("purging loader %s from constraint for name %s",
-                     probe->loader_data(n)->loader_name_and_id(),
-                     name->as_C_string());
-          }
-          probe->remove_loader_at(n);
-
-          if (lt.is_enabled()) {
-            ResourceMark rm;
-            lt.print("new loader list:");
-            for (int i = 0; i < probe->num_loaders(); i++) {
-              lt.print("    [%d]: %s", i,
-                            probe->loader_data(i)->loader_name_and_id());
+        set.remove_constraint(name, constraint);
+      } else {
+        // The class is alive or null but some of of the loaders may not be.
+        // Remove entries no longer alive from loader array
+        for (int n = constraint->num_loaders() - 1; n >= 0; n--) {
+          if (constraint->loader_data(n)->is_unloading()) {
+            if (lt.is_enabled()) {
+              ResourceMark rm;
+              lt.print("purging loader %s from constraint for name %s",
+                       constraint->loader_data(n)->loader_name_and_id(),
+                       name->as_C_string());
+            }
+            constraint->remove_loader_at(n);
+
+            if (lt.is_enabled()) {
+              ResourceMark rm;
+              lt.print("new loader list:");
+              for (int i = 0; i < constraint->num_loaders(); i++) {
+                lt.print("    [%d]: %s", i,
+                              constraint->loader_data(i)->loader_name_and_id());
+              }
             }
           }
         }
-      }
-      // Check whether the set should be purged
-      if (probe->num_loaders() < 2) {
-        if (lt.is_enabled()) {
-          ResourceMark rm;
-          lt.print("purging complete constraint for name %s",
-                   name->as_C_string());
+        // Check whether the set should be purged
+        // Only one loader doesn't enforce a constraint.
```

Comment:
So just to be clear, if we then have a second loader for a class of this name, then that would cause a constraint to be recreated, which would include both loaders?

Reply:
Yes. This is what it does.  I had to figure this out with a bunch of asserts...